src/net.cpp

```diff
@@ -1967,7 +1972,18 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
             CAddrInfo addr = addrman.SelectTriedCollision();
 
             // SelectTriedCollision returns an invalid address if it is empty.
-            if (!fFeeler || !addr.IsValid()) {
+            bool test_before_evict = fFeeler && addr.IsValid();
```

Comment:
This logic is already very convoluted, and I think this change makes it even more difficult to follow. The mainline case (connecting to a new outbound beer), currently calls `SelectedTriedCollision()`, then sets `test_before_evict` to false, and then throws away the return value from `SelectedTriedCollision()` in the `if (!test_before_evict)` block below.

The logic for feeler connections and non-feeler connections has diverged to the point where it'd make this code a lot easier to follow if it were refactored to a GetFeelerConnectionAddress() and GetOutboundConnectionAddress() function. For now, I think the following is easier to read:

```
            if (fFeeler) {
                // First, try to get a tried table collision address. This returns
                // an empty (invalid) address if there are no collisions to try.
                addr = addrman.SelectTriedCollision();

                if (!addr.IsValid()) {
                    // No tried table collisions. Select a new table address
                    // for our feeler.
                    addr = addrman.Select(true);
                } else if (AlreadyConnectedToPeer(addr)) {
                    // If test-before-evict logic would have us connect to a
                    // peer that we're already connected to, just mark that
                    // address as Good(). We won't be able to initiate the
                    // connection anyway, so this avoids inadvertently evicting
                    // a currently-connected peer.
                    addrman.Good(addr);
                    addr = addrman.Select(true);
                }
            } else {
                // Not a feeler
                addr = addrman.Select();
            }
```

Reply:
I'm indifferent, so taking your patch.